stats/src/main/java/io/airlift/stats/JmxGcMonitor.java

```diff
@@ -146,6 +153,13 @@ private synchronized void onNotification(Notification notification)
                         info.getDurationMs(),
                         info.getBeforeGcTotal(),
                         info.getAfterGcTotal());
+                if (info.getBeforeGcTotal().toBytes() / heapMemoryLimit > .8) {
+                    // Only warn of GC issues when the Heap is close to the limit
+                    double percentRecovered = 1 - info.getBeforeGcTotal().toBytes() / info.getAfterGcTotal().toBytes();
+                    if (percentRecovered < .1) {
+                        log.warn("Major GC only recovered %s% of heap memory", percentRecovered);
```

Comment:
My memory is when you are in the case that you are only freeing up a few percent memory, the GC basically runs constantly. We may want to rate limit the messages to not spam the logs.

Reply:
yeah:

```
private final RateLimiter rateLimiter = RateLimiter.create(0.1);

if (!rateLimiter.tryAcquire()) {
    return;
}
```

Reply:
Alternatively, we could replace the info log with a warn log and report as usual.

Reply:
As it is, we already spam the log when we get into a GC loop. It is just that we don't make progress.

Reply:
this will generate noise in the logs. rate limiting is a better way and we use it already in other places (see: HTTP client diagnostics)

Reply:
Yes. When I investigated an issue with a coordinator that would become unresponsive, there were lots of misleading logs but the GC logs showed that the VM was effectively in a GC loop. Unfortunately, whenever we saw this, it was usually too late.